micromanager_gui/main_window.py

```diff
@@ -110,36 +126,128 @@ def cleanup():
 
         self._add_menu()
 
-    def _add_menu(self):
+        # add minmax dockwidget
+        self.viewer.window.add_dock_widget(self.minmax, name="MinMax", area="left")
+
+    def _add_dock_widget(
+        self, widget: QWidget, name: str, floating: bool = False
+    ) -> QtViewerDockWidget:
+        dock_wdg = self.viewer.window.add_dock_widget(
+            widget,
+            name=name,
+            area="right",
+            allowed_areas=["left", "right"],
+        )
+        dock_wdg.setFloating(floating)
+        dock_wdg._close_btn = False
+        return dock_wdg
+
+    def _add_menu(self) -> None:
+        # TODO: this will be removed in the next PR we will use QtoolBar
         w = getattr(self.viewer, "__wrapped__", self.viewer).window  # don't do this.
         self._menu = QMenu("&Micro-Manager", w._qt_window)
 
-        action = self._menu.addAction("Device Property Browser...")
-        action.triggered.connect(self._show_prop_browser)
+        dp = self._menu.addAction("Device Property Browser")
+        dp.triggered.connect(self._show_dock_widget)
+
+        gp = self._menu.addAction("Groups and Presets")
+        gp.triggered.connect(self._show_dock_widget)
+
+        il = self._menu.addAction("Illumination Control")
+        il.triggered.connect(self._show_dock_widget)
+
+        st = self._menu.addAction("Stages Control")
+        st.triggered.connect(self._show_dock_widget)
+
+        cm = self._menu.addAction("Camera ROI")
+        cm.triggered.connect(self._show_dock_widget)
+
+        px = self._menu.addAction("Pixel Size")
+        px.triggered.connect(self._show_dock_widget)
+
+        m = self._menu.addAction("MDA")
+        m.triggered.connect(self._show_dock_widget)
+
+        e = self._menu.addAction("Sample Explorer")
+        e.triggered.connect(self._show_dock_widget)
 
         bar = w._qt_window.menuBar()
         bar.insertMenu(list(bar.actions())[-1], self._menu)
 
-    def _show_prop_browser(self):
-        if not hasattr(self, "_prop_browser"):
-            self._prop_browser = PropertyBrowser(mmcore=self._mmc, parent=self)
-        self._prop_browser.show()
-        self._prop_browser.raise_()
-
-    def _on_system_cfg_loaded(self):
-        if len(self._mmc.getLoadedDevices()) > 1:
-            self._set_enabled(True)
+    def _show_dock_widget(self) -> None:
+        """Method to show the selected QtViewerDockWidget."""
+        # TODO: temporary suolution, this will be
+        # changed in the next PR where the dockwidgets
+        # will be called by QPushButtons in a Qtoolbar
+
+        text = self.sender().text()
+
+        if "Browser" in text:
+            if not hasattr(self, "prop_browser_dock_wdg"):
+                self.prop_browser_dock_wdg = self._add_dock_widget(
```

Comment:
not a big fan of this general `if not hasattr` pattern.  It makes it very hard to see (in the `__init__` for example) what this `MainWindow` class *might* have.  moreover, all of these are added as public attributes, making it very tempting for someone to access them... which is almost certain to lead to an `AttributeError` given that they are lazily added

Reply:
also, the matching up of strings from the action names to the dock widget that gets added is too procedural.  I would expect to see a mapping somewhere of all the valid keys and their corresponding widgets:

```python
        WIDGETS = {
            "Device Property Browser": self.prop_browser,
            "Groups and Presets": self.group_preset_table_wdg,
            "Illumination Control": self.illumination,
            "Stages Control": self.stages,
        }
```

after which `_add_menus` could have just been:
```python
for name in WIDGETS:
    action = self._menu.addAction("Groups and Presets")
    action.triggered.connect(self._show_dock_widget)
```

and all of `_show_dock_widget` could just be:

```python
text = self.sender().text()
wdg = self._add_dock_widget(WIDGETS[text], text, floating=True)
wdg.show()
wdg.raise_()
```

(and use a dictionary somewhere to remember what widgets already have dock widgets

Reply:
what about adding them at strtup but hidden?

```
def __init__(self, viewer: napari.viewer.Viewer, remote=False):
        super().__init__()
        ...
        self.camera_roi_dock_wdg = self._add_dock_widget(
                   self.cam_roi, "Camera ROI", floating=True, hide=True
        )
        ...
```
and then 
```
def _show_dock_widget(self) -> None:
       """Method to show the selected QtViewerDockWidget."""
       text = self.sender().text()

       if "Camera" in text:
           self.camera_roi_dock_wdg.show()
           self.camera_roi_dock_wdg.raise_()
```

Reply:
maybe ... but should check whether there's any noticeable difference in load time.  And I would still probably use a dictionary rather than attributes, since it's much easier to extend as you add more things

Reply:
something like that?
```
def __init__(self, viewer: napari.viewer.Viewer, remote=False):
        super().__init__()
        ...
        self.camera_roi_dock_wdg = self._add_dock_widget(
                   self.cam_roi, "Camera ROI", floating=True, hide=True
        )
  
        self.dock_widgets = {
            "Camera ROI": self.camera_roi_dock_wdg,
            ...
            ...
        }
        ...
```
and
```
def _show_dock_widget(self) -> None:
        """Method to show the selected QtViewerDockWidget."""
        wdg = self.dock_widgets[self.sender().text()]
        wdg.show()
        wdg.raise_()
```

Reply:
ok I did't see the comment above, sorry